System/Atom.cpp

```diff
@@ -0,0 +1,1060 @@
+				}
+
+				///////////////////////////////////////////////////////////////////////////////////////////////////
+				// Apply Collision Responses
+
+				// If we hit anything, and are about to start a new segment instead of a step, apply the collision response effects to the owning MO.
+				if ((hit[X] || hit[Y]) && !m_LastHit.Terminate[HITOR]) {
+					// Calculate the progress made on this segment before hitting something.
+					// We count the hitting step made if it resulted in a terrain sink, because the Atoms weren't stepped back out of intersection.
+					//segProgress = static_cast<float>(domSteps + sinkHit) / static_cast<float>(delta[dom]);
+					segProgress = (static_cast<float>(domSteps + static_cast<int>(sinkHit)) < delta[dom]) ? (static_cast<float>(domSteps + static_cast<int>(sinkHit)) / std::fabs(static_cast<float>(segTraj[dom]))) : 1.0F;
+
+					// Now calculate the total time left to travel, according to the progress made.
+					timeLeft -= timeLeft * segProgress;
+
+					// Move position forward to the hit position.
+					//position += segTraj * segProgress;
+					// Only move the dom forward by int domSteps, so we don't cross into a pixel too far
+					position[dom] += (domSteps + static_cast<int>(sinkHit)) * increment[dom];
+
+					// Move the submissive direction forward by as many int steps, or the full float segTraj if all sub-steps are clear
+					if ((subSteps + static_cast<int>(subStepped && sinkHit)) < delta[sub]) {
+						position[sub] += (subSteps + static_cast<int>(subStepped && sinkHit)) * increment[sub];
+					} else {
+						position[sub] += segTraj[sub];
+					}
+
+					Vector testPos = position - m_Offset;
+
+					didWrap = g_SceneMan.WrapPosition(testPos) || didWrap;
+
+					// Apply the collision response acceleration to the linear velocity of the owner MO.
+					velocity += hitAccel;
+				}
+			}
+		} while ((hit[X] || hit[Y]) && /* !segTraj.GetFloored().IsZero() && */ hitCount < 100 && !m_LastHit.Terminate[HITOR]);
+
+		//RTEAssert(hitCount < 100, "Atom travel resulted in more than 100 segments!!");
+
+		// Draw the trail
+		if (g_TimerMan.DrawnSimUpdate() && m_TrailLength) {
+			int length = m_TrailLength /* + 3 * PosRand()*/;
+			for (int i = trailPoints.size() - MIN(length, trailPoints.size()); i < trailPoints.size(); ++i) {
+				putpixel(trailBitmap, trailPoints[i].first, trailPoints[i].second, m_TrailColor.GetIndex());
+			}
+		}
+
+		// Unlock all bitmaps involved.
+		//if (m_TrailLength) { trailBitmap->UnLock(); }
+		if (!scenePreLocked) { g_SceneMan.UnlockScene(); }
+
+		// Extract Atom offset.
+		position -= m_Offset;
+
+		// Travel along the remaining segTraj.
+		if (!(hit[X] || hit[Y]) && autoTravel) { position += segTraj; }
+
+		didWrap = g_SceneMan.WrapPosition(position) || didWrap;
+
+		ClearMOIDIgnoreList();
+
+		return hitCount;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void HitData::Clear() {
```

Comment:
Seems weird to have HitData defined at the top of the header but at the bottom of the cpp file.

Reply:
I just wanted to get it out of the way because you never need to look at it.

Reply:
Hmmm, fair enough. I feel like it'll throw one of us off briefly in future, but ideally no one will have to look at anything in Atom much, so it's not a big deal either way.